gcc: Use `configurePlatforms` to control --build, --host, and --target

Ericson2314 opened the pull request:
###### Motivation for this change

A step in consolidating the configure flags logic within these derivations.

###### Things done

There's quite a few combination of platforms to try. I don't expect the cross ones to work until https://github.com/NixOS/nixpkgs/pull/26805 but the native ones can be tested. @dezgeg let me know how/whether to reproduce the ARM issue for binutils, as I suspect it's a general autoconf one and might affect GCC too.

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built stdenv on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ericson2314 committed:
gcc: Use `configurePlatforms` to control --build, --host, and --target

dezgeg commented:
I threw `nix-build -A stdenv` for this on `armv7l-linux` to try it out.

Though the same question as with binutils remain, what's the concrete benefit of passing these flags when not cross compiling?

Ericson2314 commented:
I made an attempt at laying the benefits in the binutils thread https://github.com/NixOS/nixpkgs/pull/28374#issuecomment-323598869

Raspberry Pis are `armv7l`? Then I might just be able to test this out.

dezgeg commented:
Seems like it failed: https://gist.github.com/dezgeg/8d155569d7be06a846a734e8fb7123e9. Mysterious indeed...

* Raspberry Pi 1 is `armv6l-linux` (also GCC builds take like 2 days, wouldn't recommend that experience).
* Raspberry Pi 2 is `armv7l-linux`.
* Raspberry Pi 3 can be either `aarch64-linux` or `armv7l-linux`. Though unfortunately the patches for booting the mainline kernel in ARMv7 mode aren't in a released kernel yet, I think. Have to check that.

Ericson2314 commented:
Ok, then I'll do the same arm special case as binutils.

Ericson2314 commented:
Then doing the same arm workaround done for binutils should be fine.

Ericson2314 committed:
gcc: Add `configurePlatforms` exception for Arm 32

...just as we did for binutils. When the underlying issue is resolved
(probably with a configure script patch or lib/systems/parse.nix
change), this should be reverted.

Ericson2314 commented:
I added the arm 32-bit fallback.
